### DISCUSSION TITLE
Define action on canvas

### DISCUSSION THREAD

**Discussion by cool-RR** [saulpw/visidata#1690] [Q&A] [answered]
I'm trying to define an action on a canvas:

```
@Sheet.api
def toggle_plot_asterisk(canvas: Canvas) -> None:
    canvas.options.disp_canvas_charset = '*'

Canvas.addCommand('*', 'toggle-plot-asterisk', 'toggle_plot_asterisk()')
```

However, when I run it on a plot I get this error: `TypeError: toggle_plot_asterisk() missing 1 required positional argument: 'sheet'`

Why? A similar pattern works well for `BaseSheet`.

**Comment by saulpw** [comment, marked as answer]
`Sheet` is an alias for `TableSheet`, which only handles all sheets with rows and columns--a `Canvas` is actually not a `TableSheet`.  `BaseSheet` is the ultimate parent sheet class for all "sheets" that get pushed, including `Canvas`es.

If you decorate `toggle_plot_asterisk` with `@Canvas.api` instead, it should work.  (Note that your first parameter is a Canvas, not a Sheet).  The error is confusing, but is because all functions are implicitly globals exposed in `.visidatarc`, including functions decorated with the monkey-patching `.api`!  The command executor will try the `vd` and `sheet` contexts first, but if an identifier isn't on there, it will look in globals; and then fail in this way.

Other ways to get different results (I mention these for completeness but the above fix is the most correct):

- using `sheet.toggle_plot_asterisk()` in the execstr; this won't fix the problem, but explicitly using `sheet` will give you the more correct error that the method isn't on the Canvas;
- using `@BaseSheet.api`; this will work fine and isn't a problem, especially for individual use like in .visidatarc, but is overly broad as it makes the method available on all sheets;
- omitting the `@Sheet.api` decorator completely and using `toggle_plot_asterisk(sheet)` (this will work in .visidatarc for the reason mentioned above, but not in a plugin, unless added explicitly with `addGlobals()`

**Comment by cool-RR** [comment]
Great, that works, thank you. 

Another question: Here's how I wrote the code:

```
@Canvas.api
def toggle_plot_asterisk(canvas: Canvas) -> None:
    if canvas.options.disp_canvas_charset == '*':
        canvas.options.disp_canvas_charset = DEFAULT_DISP_CANVAS_CHARSET
    else:
        canvas.options.disp_canvas_charset = '*'
```

I defined `DEFAULT_DISP_CANVAS_CHARSET` in my code because I couldn't figure out how to get that default value from VisiData. Is that possible?

**Reply by saulpw** [reply]
That seems fine.  You could also do:

```
DEFAULT_DISP_CANVAS_CHARSET = vd.options.disp_canvas_charset
```

at the top-level of your code.

**Reply by cool-RR** [reply]
That's better than my version, thank you. I still think it'll be nice if default values will always be accessible in VisiData code even if they were overridden, but that's not critical.

**Reply by cool-RR** [reply]
By the way, my stupid hack was to define it this way:

```
DEFAULT_DISP_CANVAS_CHARSET = lzma.decompress(base64.b64decode('/Td6WFoAAATm1rRGAgAhARYAAAB0L+Wj4AL/APVdAHEoDBAEgdA/tto4kuI0/tYWZYs4kpFqcYNhVWuCYAjoA+51kP/qzYvJ+dlmd2aAk+rB9SBN+ozdt+cviLig0iJrNKjuKVlY2yphXBjNq+09pIXFlJGX+4c+YPaqQchDSHV9MT3qccRZHQZj0KNLafy2c5Of7VNRVe8cx3rtRG9bjKSr8GI20xsQNa4FLs9cbNRFJPaTnhpoH3vDwWQ2mB0KzEAQyz/stMAp/jfosrpfLSXXLsL2bShczRUOX00NqkELz09PltBa93A39jJ4lASWUwrWlTLJ5yGHKtTUCIRFmNg7TPiZs5CYYbj9GEN8qkW1EHR2AAAAALVhmWb/bhXDAAGRAoAGAADSiHuGscRn+wIAAAAABFla')).decode()
```

Just so the file will be ascii and the line would be as short as possible :laughing:
